### PR TITLE
fix(cons messages): include in header hash for fraud protection 

### PR DESCRIPTION
**Pull request by danwt**
# PR Standards

## Opening a pull request should be able to meet the following requirements

--

PR naming convention: https://hackmd.io/@nZpxHZ0CT7O5ngTp0TP9mg/HJP_jrm7A

---

Close #XXX

<-- Briefly describe the content of this pull request -->

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

**Review by srene — COMMENTED** [review]
i tested locally with a sequencer + fullnode and only upsert msgs without upgrade, but seems to work fine.

File: types/errors.go [hunk 1]
```diff
@@ -390,6 +390,28 @@ func (e ErrInvalidHeaderDataHashFraud) Unwrap() error {
...
+
+func NewErrInvalidDymHeaderFraud(block *Block) error {
+	return &ErrInvalidDymHeaderFraud{
+		block,
+	}
+}
+
+func (e ErrInvalidDymHeaderFraud) Error() string {
+	return fmt.Sprintf(
+		"possible fraud: invalid dym header: height : %d",
```
[the remark this answers is not in the document]
Comment (danwt) [reply types/errors.go:406]: it's in the block
fraud handler should do something with that

[the remark this answers is not in the document]
Comment (danwt) [reply types/errors.go:406]: having the errors spit out all this info isn't the right pattern

File: block/executor.go [hunk 2]
```diff
@@ -129,7 +129,7 @@ func (e *Executor) InitChain(genesis *tmtypes.GenesisDoc, genesisChecksum string
 			Version: &tmproto.VersionParams{
 				AppVersion: params.Version.AppVersion,
 			},
-		}, Validators: valUpdates,
+		}, Validators:   valUpdates,
```
[the remark this answers is not in the document]
Comment (danwt) [reply block/executor.go]: You know there are about a dozen files which are not gofumpted?

File: types/errors.go [hunk 3]
```diff
@@ -19,6 +19,7 @@ var (
 	ErrEmptyBlock            = errors.New("block has no transactions and is not allowed to be empty")
 	ErrInvalidBlockHeight    = errors.New("invalid block height")
 	ErrInvalidHeaderDataHash = errors.New("header not matching block data")
+	ErrInvalidDymHeader      = gerrc.ErrInvalidArgument.Wrap("dym header")
```
[the remark this answers is not in the document]
Comment (danwt) [reply types/errors.go:22]: Because it's a type validation

**Review by omritoptix — APPROVED** [review]

File: block/executor_test.go [hunk 4]
```diff
@@ -86,8 +87,40 @@ func TestCreateBlock(t *testing.T) {
...
+	require.NoError(err)
+	require.NoError(b3.ValidateBasic())
+
+	// only to proto
+	require.NoError(b1.ValidateBasic())
+	b1p3 := b1.ToProto()
+	var b4 types.Block
+	err = b4.FromProto(b1p3)
+	require.NoError(err)
+	require.NoError(b4.ValidateBasic())
```
Comment (danwt) [inline block/executor_test.go:122]: ugly but just wanted to check 10x

**Review by omritoptix — APPROVED** [review]

**Pull request merged**